docs/en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_ReactJS_Redux_and_GraphQL.md

```diff
@@ -0,0 +1,666 @@
+
+```js
+{
+	"user": {
+		"name": "Test user",
+		"email": "me@example.com",
+		"blogPosts": [
+			{
+				"title": "How to be awesome at GraphQL",
+				"comments": [
+					{
+						"author": "Uncle Cheese",
+						"comment": "Nice stuff, bro"
+					}
+				]
+			}
+		]
+	}
+}
+```
+
+On its own, GraphQL offers nothing functional, as it's just a query language. You still need a service that will invoke queries and map their results to UI. For that, SilverStripe uses an implementation of [Apollo](http://dev.apollodata.com/) that works with React.
+
+## For more information
+
+This documentation will stop short of explaining React, Redux, and GraphQL/Apollo in-depth, as there is much better
+documentation available all over the web. We recommend:
+* [The Official React Tutorial](https://facebook.github.io/react/tutorial/tutorial.html)
+* [Build With React](http://buildwithreact.com/tutorial)
+* [Getting Started with Redux](https://egghead.io/courses/getting-started-with-redux)
+* [The React Apollo docs](http://dev.apollodata.com/react/)
+
+# The Injector API
+
+Much like customisation is done in the backend through [Injector](../../extending/injector), the client-side framework has its own implementation of dependency injection known as `Injector`. Using Injector, you can register new services, and transform existing services.
+
+Injector is broken up into three sub-APIs:
+* `Injector.react` for React UI components
+* `Injector.reducer` for Redux state management
+* `Injector.form` for forms rendered via `FormSchema`.
+
+The frontend Injector works a bit differently than its backend counterpart. Instead of _overriding_ a service with your own implementation, you _enhance_ an existing service with your own concerns. This pattern is known as [middleware](https://en.wikipedia.org/wiki/Middleware).
+
+Middleware works a lot like a decorator. It doesn't alter the original API of the service,
+but it can augment it with new features and concerns. This has the inherent advantage of allowing all thidparty code to have an influence over the behaviour, state, and UI of a component.
+
+## A simple middleware example
+
+Let's say you have an application that features error logging. By default, the error logging service simply outputs to `console.error`. But you want to customise it to send errors to a thirdparty service. For this, you could use middleware to augment the default functionality of the logger.
+
+_LoggingService.js_
+```js
+(error) => console.error(error);
+```
+
+Now, let's add some middleware to that service. The signature of middleware is: 
+```js
+(next) => (args) => next(args)
+```
+Where `next()` is the next customisation in the "chain" of middleware. Before invoking the next implementation, you can add whatever customisations you need. Here's how we would use middleware to enhance `LoggingService`.
+
+```js
+(LoggingService) => (error) => {
```

Comment:
It's a bit weird to write invalid JS here. Which I guess you've done because `Injector.transform()` hasn't been introduced. How about moving this middleware section underneath "registering new services with Injector", and folding the middleware intro into "Transforming services using middleware"? Alternatively, you'll need to mention why this JS is invalid, and that we'll complete it later.